### ISSUE TITLE
getTriples for AnnotationCollection Is not Displaying Description 

### ISSUE THREAD

**Issue by svaibhav19** [anno4j/anno4j#149]
hello,

Issue 1:
I have been using the AnnotationCollection and AnnotationPage for storing list of annotation.
The storing of multiple annotation works fine, but 
while displaying the RDF-XML representation (getTriples();) it only shows the list of items, rather that the content of the items, which I was expecting. 
An example is shown below,
<?xml version="1.0" encoding="UTF-8"?>
<rdf:RDF
	xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">

<rdf:Description rdf:about="urn:anno4j:ab059895-b048-4c44-97d3-c3867751bf6e">
	<rdf:type rdf:resource="http://www.w3.org/ns/activitystreams#OrderedCollection"/>
	<totelItems xmlns="http://www.w3.org/ns/activitystreams#" rdf:datatype="http://www.w3.org/2001/XMLSchema#int">4</totelItems>
	<first xmlns="http://www.w3.org/ns/activitystreams#" rdf:resource="urn:anno4j:f53cbc69-42ab-4e5d-9984-a35bec508966"/>
</rdf:Description>

</rdf:RDF>
From the above XML the ID: urn:anno4j:f53cbc69-42ab-4e5d-9984-a35bec508966 is missing its details of list of annoations.
When i tried printing AnnotationPage i could see only the items with ID but again the Description for each item is missing.
Whlie refereing to code, I saw the getTriple Method has not been overiden in this AnnotationCollectionSupport and AnnotationPageSupport classes, 
which may be used for printing items and related Description of it. However, I am not sure whether this is the issue or not. 
This issue might be similar to open issue 139(Interface for merging Choice and Composite Classes?)
Which has been solved (Issue 139) for the snapshot version by Emanuel/Manu.

Issue 2:

I have created Annoatation using Anno4j and generated the JSON-LD representation.
Then I  created an ObjectParser and given the input of generated JSON-LD.
List<Annotation> annotations = objectParser.parse(annotation.getTriples(RDFFormat.JSONLD), new URL("http://www.w3.org/"), RDFFormat.JSONLD);
But i am not able to get the respective RDF representation.
Generated JSON-LD is quite large to share, so i have not added to this issue but can share if needed.

For the above stated two issues, I am not sure whether they are issues or errors from lack of my understanding, or they are due to the missing implementation in the library.
It will be very helpfull if you could shed some light on these issues.

**Comment by berndl** [comment]
Hi @svaibhav19 !

Finally I had some time to look into your issues. You are totally right, the getTriples() for the AnnotationPage and AnnotationCollection were missing.
Implementations were added to those and I am currently pushing the changes to the develop branch. You should be able to test those in some minutes.
I was not quite happy with the basic behaviour for the classes, as in RDF you would have to do many things by hand, e.g. adding a page to a collection does not imply that the "partOf" relationship is also set. I tried to add some of this, too, but I am currently limited in time, so this might not be finished completely.

Please use the updated code and check if the getTriples fits your requirements. I will hopefully soon have also time to check your second issue.

Best regards,
Manu

**Comment by berndl** [comment]
Okay, I could look also in the second issue, but I am not quite sure where your issue lies without seeing your code. I agree that the JSON-LD will be quite big, but maybe you can support the code snippet in which you use the ObjectParser?

I also created a minimal example in our test suite, see https://github.com/anno4j/anno4j/blob/develop/anno4j-core/src/test/java/com/github/anno4j/io/InputOutputTest.java#L95. Maybe that one can help you as well.

Please inform me if the issues still persist!

Best,
Manu

**Comment by svaibhav19** [comment]
Hi @berndl ,

First of all I would like to thankyou for your response and sorry for replying late.
I have tested the AnnotationCollection and AnnotationPage modification, it is working fine and fits to the requirements.
Thankyou, for your valuable time and efforts.

Regarding The second issue, i have checked the test but my issue still remains the same.
Please find the below code snippet,

Anno4j anno4j = new Anno4j();
Annotation annotation = anno4j.createObject(Annotation.class);

SpecificResource specificResource = anno4j.createObject(SpecificResource.class);
ResourceObject source = anno4j.createObject(ResourceObject.class);
specificResource.setSource(source);

TextualBody textualBody = anno4j.createObject(TextualBody.class);
textualBody.setValue("someText");

annotation.addTarget(specificResource);
annotation.addBody(textualBody);

// Create JSONLD of the Annotation
String jsonld = annotation.getTriples(RDFFormat.JSONLD);
System.out.println("Output:\n"+jsonld);
System.out.println("\n\n---------------\n\n");
// Parse the JSONLD String
ObjectParser parser = new ObjectParser();
List<Annotation> parsed = parser.parse(jsonld, new URL("http://example.com/"), RDFFormat.JSONLD);
for (Annotation annotation2 : parsed) {
	System.out.println("No content to Show :"+annotation2.getTriples(RDFFormat.RDFXML));
	System.out.println(".getResourceAsString() this works:"+annotation2.getResourceAsString());
}

I tried to print all the annotation in RDFXML format, the getResourceAsString works completely fine 
but when I try to print the getTriples form the parsed annotation (annotation2) it is empty.
Is it correct what I am trying to do or I have the wrong implementation?

Thank you once again, this Anno4j library has helped a lot for the implementation of the Web Annotation Data Model.

Thanks And Best Regards
Vaibhav Sakharkar

**Comment by berndl** [comment]
Hello again @svaibhav19 !

First of all I am really happy that you use our library and that I can do something to make it better for you to use :) Thanks for committing your issue!

I finally found the solution, which is kind of a tricky one, as we had another issue before ( #131 ) with the ObjectParser.

But first to the solution: I changed the .parse() method of the ObjectParser a little bit, as i added a "clear" boolean parameter. In your case, we need to set it to false, so the triples about your parsed Annotations will persist. This way, they can be printed via the .getTriples() method afterwards.
I created a Gist with the code you supported: [Gist with adapted code from svaibhav19](https://gist.github.com/berndl/1fc9a12ac566fe14b35000f6728e2ef2) (I hope this works, i never used the Gist function :)) 

Just be careful with that: the issue I mentioned before had the problem of bad performance with the ObjectParser when to many Annotations are parsed at a time. This is why we cleared the locally parsed Annotations after a call to .parse().
To counteract this with the "clear" parameter set to false, you should "reset" the ObjectParser from time to time, for example with the call to its .shutdown() method. I added this also to the Gist snippet. But the call to .getTriples() must be done BEFORE the call to shutdown. Otherwise the triples are lost like in the original issue.

I pushed the code already to our development branch, so it should be accessible soon with the latest SNAPSHOT version.

I hope I could help you with this!

Please come back if any problems still persist.

Best regards,
Manu

PS: There is also another solution to this. I added it as a comment to the test method that I linked last time: [Link to InputOutputTest Class](https://github.com/anno4j/anno4j/blob/develop/anno4j-core/src/test/java/com/github/anno4j/io/InputOutputTest.java#L134). What one could also do is persist the parsed Annotations to your stable Anno4j instance (not the local one used by the ObjectParser), and then query them in the standard way. This however seemed pretty inconvenient for me :)

**Comment by fischmat** [comment]
Hi Manu,

habe das Tool jetzt soweit fertig, dass man Source-Files reingibt und JARs rausbekommt.
Das Ganze ist jetzt ein eigenes Maven Submodul und sollte beim Deployment als ausführbares JAR mit gebaut werden (lt. Stackoverflow 😄).
Momentan habe ich es gerade auf dem Server laufen und hoffe, dass ich bald die Proxies fürs ECRM habe.


Viele Grüße und schönes Wochenende,

Matthias

**Comment by svaibhav19** [comment]
Hi @berndl ,

Thank you so much for the updates, i have tested the code and its working.

But as i have taken the updates i started running with some other problems.
The same problem that i had earlier regarding printing the gettripples issue 139 (similar but with minor difference).
I am trying to add almost 69+ body items to the CHOICE with their properties, and when i try to display using gettripples
it is displaying the ID associated to it but no details are getting displayed.
Whereas, when i try to add only few items to CHOICE ( I have tested for 3 to 30 body items ) it works fine but when it goes beyond 30 items,
the behaviour gets changed and it does not display description of it.
This issue i faced after the new update, earlier it was working fine.

Also i have one more question does the Anno4j can generate the output as per
https://www.w3.org/TR/annotation-model/#serialization-of-the-model anno.jsonld format.

Thanks & Best Regards
Vaibhav Sakharkar

Sorry , i have accidentally closed the issue. Extremely sorry for the trouble. do it is fine if it is reopen or start again with the new thread.

**Comment by svaibhav19** [comment]
Hi @berndl ,

I have tried adding multiple choice items but up to certain limit the adding choice item is working perfect, but as i keep on adding several items to choice its behavior gets changed and  stop displaying all the description of the same. Link 1: https://github.com/svaibhav19/Thesis/blob/1d1f24b7c1d8a58738b64d231edd328977fdf7ee/AnnotationWADM_v1.0/resources/pageAnnotation.xml
Link 2: https://github.com/svaibhav19/Thesis/blob/master/AnnotationWADM_v1.0/resources/pageAnnotationfull.xml 
As i have added the comment in the previous message, please find the link 1 which is the short version and working fine and link 2 will help to identify the missing elements of choice items.
Even I tried to print all the items individually and it prints as per the requirement but when i try to print complete annotation it fails to display the items content.

Thanks & Best Regards
Vaibhav Sakharkar

**Comment by berndl** [comment]
Hi @svaibhav19 ,

it sure is ok to reopen this issue :) It still concerns your same problem.

I've found the issue. It is a weird timing thing that we encountered only once until now - now you have found a second occasions, thank you for that! :)
Some objects implement different behaviours from superclasses, and depending on execution order, one behaviour dictates how the object is printed via the getTriples method. This is why you sometimes faced a "correct" output and sometimes the wrong one.

As usual, I pushed the fix to the development branch, so it will soon be available for you to test!

What do you mean with the same output as in the WADM specification? Anno4j does support printing JSONLD when you support the "RDFFormat.JSONLD" for the getTriples method. I think the specification document does support some more syntactic sugar in the printed version, semantically everything should be the same.

Best regards,
Manu

**Comment by svaibhav19** [comment]
Hi @berndl ,

Thank you for all your support and guidance.

Thanks & Best Regards
Vaibhav Sakharkar
